src/msgpack/ArrayConverter.cs

```diff
@@ -0,0 +1,103 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+namespace TarantoolDnx.MsgPack
+{
+    internal class ArrayConverter<TArray, TElement> : ArrayConverterBase<TArray, TElement>
+        where TArray : IList<TElement>
+    {
+        private static readonly bool IsSingleDimensionArray;
+
+        static ArrayConverter()
+        {
+            var type = typeof(TArray);
+            IsSingleDimensionArray = type.IsArray && type.GetArrayRank() == 1 && type.GetElementType() == typeof(TElement);
+        }
+
+        public override void Write(TArray value, Stream stream, MsgPackSettings settings)
+        {
+            if (value == null)
+            {
+                settings.NullConverter.Write(value, stream, settings);
+                return;
+            }
+
+            WriteArrayHeaderAndLength(value.Count, stream);
+            var elementConverter = settings.GetConverter<TElement>();
```

Comment:
@Ivan-Razhnou it's context job to cache converters. Couple of hash lookups per write call is not a problem. And converter is per-call specific. You can override it for next call via context and use same converter for array.